### PR TITLE
Teads AWS Model

### PR DESCRIPTION
**Pull request by gnanakeethan**
### Types of changes
- [ ] Enhancement (project structure, spelling, grammar, formatting)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### A description of the changes proposed in the Pull Request
Implementation for Teads' Model for AWS Instances Carbon Footprint Calculations.

### CODE REVIEW

**Review by jawache — CHANGES_REQUESTED** [review]
Requested a few changes thanks

File: src/lib/teads-aws/index.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    const y: number[] = [
+      this.computeInstances[this.instanceType].consumption.idle ?? 0,
+      this.computeInstances[this.instanceType].consumption.tenPercent ?? 0,
+      this.computeInstances[this.instanceType].consumption.fiftyPercent ?? 0,
+      this.computeInstances[this.instanceType].consumption.hundredPercent ?? 0,
+    ];
+
+    const spline = new Spline(x, y);
+
+    const wattage = spline.at(cpu);
```
[the remark this answers is not in the document]
Comment (gnanakeethan) [reply src/lib/teads-aws/index.ts:143]: @jawache It becomes a complex interpretation of the term interpolation. 

CCF does not use TEADS curve data at all. They are summing it to the CPU Chip Architecture and they use the min max to compute interpolation. 

We can do interpolation between the 10 & 50 for 20%. 

I will add the configuration params & options.

File: src/lib/teads-aws/index.ts [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+      let max = 1;
+      if (cpu > 10 && cpu <= 50) {
+        min = 1;
+        max = 2;
+      }
+      if (cpu > 50 && cpu <= 100) {
+        min = 2;
+        max = 3;
+      }
+      wattage = y[0] + ((y[max] - y[min]) / (x[max] - x[min])) * cpu;
```
Comment (jawache) [inline src/lib/teads-aws/index.ts:201]: Hmm are you sure it's not just y[min] + ..... ?

**Review by jawache — COMMENTED** [review]
@gnanakeethan added a few more comments.

File: src/lib/teads-aws/index.ts (hunk 2 above)
Comment (gnanakeethan) [reply src/lib/teads-aws/index.ts:201]: @jawache I am implementing an improved mechanism. I will copy it over from Teads CPU.

File: src/lib/teads-aws/index.ts [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+  name: string | undefined;
+  // compute instances grouped by the provider with usage data
+  private computeInstances: {
+    [key: string]: KeyValuePair;
+  } = {};
+
+  // list of all the by Architecture
+  private instanceType = '';
+  private expectedLifespan = 4;
+  private interpolation = Interpolation.SPLINE;
```
[the remark this answers is not in the document]
Comment (gnanakeethan) [reply src/lib/teads-aws/index.ts:25]: Teads is curve based. I thought we better stick to the curve methodology. 

Main difference happens when you have less than 50% usage. 

which going to be the typical scenario for optimization cases. 

I am planning to make the curve configurable here as well.

[the remark this answers is not in the document]
Comment (jawache) [reply src/lib/teads-aws/index.ts:25]: Don't make the curve configurable, this is the teads curve and those coefficients are fixed.

Sounds like you don't have proof people are using spline and are making an assumption? If so stick to linear as the default

[the remark this answers is not in the document]
Comment (gnanakeethan) [reply src/lib/teads-aws/index.ts:25]: Sure. Switching to linear.

File: src/lib/teads-aws/index.ts [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+      if (instanceType in this.computeInstances) {
+        this.instanceType = instanceType;
+      } else {
+        throw new Error('Instance Type not supported');
+      }
+    } else if (this.instanceType === '') {
+      throw new Error('Instance Type not provided');
+    }
+
+    if ('expected_lifespan' in staticParams) {
```
Comment (jawache) [inline src/lib/teads-aws/index.ts:67]: Needs to be updated to standard param and unit.

File: src/lib/teads-aws/index.ts [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    name: string,
+    staticParams: object | undefined = undefined
+  ): Promise<IImpactModelInterface> {
+    this.name = name;
+
+    if (staticParams === undefined) {
+      throw new Error('Required Parameters not provided');
+    }
+
+    if ('instance_type' in staticParams) {
```
Comment (jawache) [inline src/lib/teads-aws/index.ts:56]: Also add to units.yml (as instance-type)

File: src/lib/teads-aws/index.ts [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+      throw new Error(
+        'Required Parameters duration,cpu,datetime not provided for observation'
+      );
+    }
+
+    //    duration is in seconds
+    const duration = observation['duration'];
+
+    //    convert cpu usage to percentage
+    const cpu = observation['cpu'];
```
Comment (jawache) [inline src/lib/teads-aws/index.ts:173]: This is now cpu-util

File: src/lib/teads-aws/index.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+   * cpu: cpu usage in percentage
+   * datetime: ISO 8601 datetime string
+   *
+   * Uses a spline method for AWS and linear interpolation for GCP and Azure
+   */
+  private calculateEnergy(observation: KeyValuePair) {
+    if (
+      !('duration' in observation) ||
+      !('cpu' in observation) ||
+      !('datetime' in observation)
```
Comment (jawache) [inline src/lib/teads-aws/index.ts:162]: This is now timestamp I believe

File: src/lib/teads-aws/index.ts [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    if (this.instanceType === '') {
+      throw new Error('Configuration is incomplete');
+    }
+
+    return observations.map((observation: KeyValuePair) => {
+      this.configure(this.name!, observation);
+      const e = this.calculateEnergy(observation);
+      const m = this.embodiedEmissions(observation);
+      observation['energy'] = e;
+      observation['embodied'] = m;
```
Comment (jawache) [inline src/lib/teads-aws/index.ts:106]: This is now embodied-carbon

**Review by jawache — CHANGES_REQUESTED** [review]
Normalizing the parameter names to the standard units

**Review by narekhovhannisyan — APPROVED** [review]

**Review by jawache — APPROVED** [review]

**Pull request merged**